### PR TITLE
feat: move to vite for bundling

### LINKED ISSUE #77: Move to vite for bundling
Currently using `microbundlecrl` which is not maintained and hard to configure.
Moving to vite for bundling will abstract some of the `rollup` config while still allowing for customization.

ref: https://jivancic.com/posts/build-a-component-library.html

### PR DESCRIPTION
**Pull request by spaenleh**
Also move to `@tanstack/react-query` and make it a peer-dependency.

closes #77

### CODE REVIEW

**Review by swouf — APPROVED** [review]
Nice job 👏  I don't have any major comments, just a few questions and remarks.

Just a general remark: Will the update to immutable 4.3.0 break the apps that will upgrade to this version of apps-query-client? Because a breaking change in the `List` type was conflicting when upgrading to immutable 4.3.0 in the apps if I remember well. It may be something to pay attention to.

File: package.json [hunk 1]
```diff
@@ -16,72 +16,76 @@
...
+    "@vitejs/plugin-react": "3.1.0",
     "eslint": "8.29.0",
     "eslint-config-prettier": "8.5.0",
     "husky": "8.0.2",
     "jest": "28.1.3",
-    "microbundle": "0.15.1",
     "nock": "13.2.9",
     "prettier": "2.8.1",
+    "react": "18.2.0",
+    "react-dom": "18.2.0",
```
Comment (dialexo) [inline package.json:72]: Same as above

File: src/hooks/postMessage.ts [hunk 2]
```diff
@@ -74,22 +74,24 @@ const configurePostMessageHooks = (_queryClient: QueryClient, queryConfig: Query
...
         reject: (reason?: unknown) => void;
       },
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
       formatResolvedValue?: (data: { payload: any; event: MessageEvent }) => A,
     ) =>
     (event: MessageEvent) => {
       try {
         const { type, payload } = JSON.parse(event.data) || {};
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
         const format = formatResolvedValue ?? ((data: { payload: any }) => data.payload);
```
Comment (dialexo) [inline src/hooks/postMessage.ts:84]: Same as above

File: test/utils.tsx [hunk 3]
```diff
@@ -1,15 +1,16 @@
...
 import { StatusCodes } from 'http-status-codes';
-import { QueryObserverBaseResult, MutationObserverResult } from 'react-query';
+import { QueryObserverBaseResult, MutationObserverResult } from '@tanstack/react-query';
 import configureHooks from '../src/hooks';
 import { Notifier, QueryClientConfig } from '../src/types';
 import { API_HOST, MOCK_APP_ORIGIN, REQUEST_METHODS } from './constants';
 import configureQueryClient from '../src/queryClient';
 
 type Args = { enableWebsocket?: boolean; notifier?: Notifier; GRAASP_APP_ID?: string | null };
+type NockRequestMethods = Lowercase<`${REQUEST_METHODS}`>;
```
Comment (dialexo) [inline test/utils.tsx:13]: I didn't know that `Lowercase<T>` was a thing :sweat_smile:

**Review by dialexo — CHANGES_REQUESTED** [review]
Thank you for your work :eagle: I've left a few questions and comments. In general, what does Vite bring to a library? I guess it's a good thing to use the same tools everywhere, but are there any pros/cons compared to using it only for frontends / bundling libraries separately with e.g. microbundle or rollup? Do we get live reloads in libraries as well?

File: tsconfig.json [hunk 4]
```diff
@@ -1,30 +1,25 @@
...
-    "noImplicitThis": true,
-    "noImplicitAny": true,
-    "strictNullChecks": true,
-    "suppressImplicitAnyIndexErrors": true,
-    "noUnusedLocals": true,
-    "noUnusedParameters": true,
+    "target": "ESNext",
+    "useDefineForClassFields": true,
+    "lib": ["DOM", "DOM.Iterable", "ESNext"],
+    "allowJs": false,
```
[the remark this answers is not in the document]
Comment (spaenleh) [reply tsconfig.json:6]: Yes this is the way 😉

File: package.json [hunk 5]
```diff
@@ -16,72 +16,76 @@
...
     "qs": "6.11.0",
-    "react-query": "4.0.0-beta.23",
-    "standard-version": "9.5.0",
-    "tsc": "2.0.4",
     "uuid": "9.0.0"
   },
   "devDependencies": {
     "@commitlint/cli": "17.3.0",
     "@commitlint/config-conventional": "17.3.0",
+    "@tanstack/react-query": "4.28.0",
```
[the remark this answers is not in the document]
Comment (spaenleh) [reply package.json:53]: Yes, this is to make it so the consumer can sync their preffered version of react-query.

File: package.json (hunk 1 above)
Comment (spaenleh) [reply package.json:72]: It is still needed that you have the packages when running i.e. the tests, so this is why they are `devdependencies`

File: test/utils.tsx (hunk 3 above)
Comment (spaenleh) [reply test/utils.tsx:13]: Well I had to add this in order to avoid the TS error related to the index signatures ...

**Comment by spaenleh** [comment]
@dialexo 
> In general, what does Vite bring to a library?

Vite uses rollup to bundle the library with __good__ defaults. So you have the advantages of rollup whie not having to setup everything yourself. 

> I guess it's a good thing to use the same tools everywhere, but are there any pros/cons compared to using it only for frontends / bundling libraries separately with e.g. microbundle or rollup? Do we get live reloads in libraries as well?

Yes you get the added benefit of using the same tools everywhere. Vite uses rollup to build your library so 👍 

Cherry on top would be to use [vitest](https://vitest.dev/) in replacement of jest.

**Review by dialexo — APPROVED** [review]

**Pull request merged**